RecoEgamma/EgammaTools/plugins/HGCalElectronFilter.cc

```diff
@@ -0,0 +1,139 @@
+// -*- C++ -*-
+//
+// Package:    EgammaTools/HGCalElectronFilter
+// Class:      HGCalElectronFilter
+//
+/**\class HGCalElectronFilter HGCalElectronFilter.cc EgammaTools/HGCalElectronFilter/plugins/HGCalElectronFilter.cc
+
+ Description: filtering of duplicate HGCAL electrons
+
+ Implementation:
+     [Notes on implementation]
+*/
+//
+// Original Author:  Florian beaudette
+//         Created:  Mon, 06 Nov 2017 21:49:54 GMT
+//
+//
+
+
+// system include files
+#include <memory>
+
+// user include files
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/StreamID.h"
+
+#include "DataFormats/Math/interface/deltaR.h"
+#include "DataFormats/EgammaCandidates/interface/GsfElectron.h"
+#include "DataFormats/EgammaCandidates/interface/GsfElectronFwd.h"
+
+//
+// class declaration
+//
+
+class HGCalElectronFilter : public edm::stream::EDProducer<> {
+   public:
+      explicit HGCalElectronFilter(const edm::ParameterSet&);
+      ~HGCalElectronFilter() override;
+
+      static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+   private:
+      void beginStream(edm::StreamID) override;
+      void produce(edm::Event&, const edm::EventSetup&) override;
+      void endStream() override;
+
+      // ----------member data ---------------------------
+      edm::EDGetTokenT<edm::View<reco::GsfElectron>> electronsToken_;
+      std::string  outputCollection_;
+      bool cleanBarrel_;
+};
+
+
+HGCalElectronFilter::HGCalElectronFilter(const edm::ParameterSet& iConfig):
+    electronsToken_(consumes<edm::View<reco::GsfElectron>>(iConfig.getParameter<edm::InputTag>("inputGsfElectrons"))),
+    outputCollection_(iConfig.getParameter<std::string>("outputCollection")),
+    cleanBarrel_(iConfig.getParameter<bool>("cleanBarrel"))
+{
+    produces<reco::GsfElectronCollection>(outputCollection_);
+}
+
+
+HGCalElectronFilter::~HGCalElectronFilter() {
+}
+
+
+//
+// member functions
+//
+
+// ------------ method called to produce the data  ------------
+void
+HGCalElectronFilter::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+
+    auto gsfElectrons_p = std::make_unique<reco::GsfElectronCollection>();
+
+    edm::Handle<edm::View<reco::GsfElectron>> electronsH;
+    iEvent.getByToken(electronsToken_, electronsH);
+
+    for(const auto& electron1 : *electronsH) {
```

Comment:
This nested loop could almost certainly be optimized further to avoid repeated comparisons. Should be considered for a followup PR. How large is the electron collection in high PU?

Reply:
Input is ~200, however I think the supercluster ref comparison should be pretty fast.
For sure we could make it `n(n+1)/2` in a future PR

Reply:
If the pT preselection is done already before, it will speed up the loop even further.

Reply:
It's 85th on the module time list, let's not worry about it for now.

Reply:
You can make it linear using a std::unordered_set on the (super)cluster refs.... Just sayin'...